packages/core/src/lib/sign-in.ts

```diff
@@ -65,3 +71,58 @@ export const signInWithEmailAndPasscode = async (
 
   await assignSignInResult(ctx, provider, id);
 };
+
+// TODO: change this after frontend is ready.
+const connectorRedirectUrl = 'https://logto.dev/callback';
+
+export const prepareSignInWithConnector = async (
+  ctx: WithUserLogContext<Context>,
+  connectorId: string,
+  state: string
+) => {
+  const connector = await getSocialConnectorInstanceById(connectorId);
+  assertThat(connector.connector?.enabled, 'connector.not_found');
+  const redirectTo = await connector.getAuthorizationUri(connectorRedirectUrl, state);
+  ctx.body = { redirectTo };
+};
+
+const getConnector = async (connectorId: string) => {
+  try {
+    return await getSocialConnectorInstanceById(connectorId);
+  } catch (error: unknown) {
+    // Throw a new error with status 422 when connector not found.
+    if (error instanceof RequestError && error.code === 'entity.not_found') {
+      throw new RequestError({
+        code: 'session.invalid_connector_id',
+        status: 422,
+        data: { connectorId },
+      });
+    }
+    throw error;
+  }
+};
+
+export const signInWithConnector = async (
+  ctx: WithUserLogContext<Context>,
+  provider: Provider,
+  { connectorId, code }: { connectorId: string; code: string }
+) => {
+  ctx.userLog.connectorId = connectorId;
+  ctx.userLog.type = UserLogType.SignInEmail;
+
+  const connector = await getConnector(connectorId);
+  const accessToken = await connector.getAccessToken(code);
+
+  const userInfo = await connector.getUserInfo(accessToken);
```

Comment:
Will any connector that supports returning both access_token and userInfo in idtoken within a single request?

Reply:
After a survey, most of connector's using OAuth will have both `getAccessToken` and `getUserInfo` API, if a connector supports returning userInfo along with accessToken, then it can catch the "userInfo" and return it when `connector.getUserInfo` is called, so we can keep current design.

Reply:
nit: can we use a more specific name such as `getUserInfoFromSocialProvider`? this method name reminds me of `getUserInfo` (from our own DB).